skrifa/src/color/mod.rs

```diff
@@ -1,6 +1,406 @@
-//! Color outlines.
+//! Drawing color glyphs.
+//!
+//! TODO(add example, including provider, getcolorpaintables)
+pub mod instance;
+pub mod transforms;
 
-// Temporary until remaining API is filled in.
-#![allow(dead_code)]
+pub(crate) mod colrv1_traversal;
 
-mod instance;
+#[cfg(test)]
+mod traversal_tests;
+
+use raw::tables::colr;
+#[cfg(test)]
+use serde::{Deserialize, Serialize};
+
+use read_fonts::{
+    tables::colr::{CompositeMode, Extend},
+    types::{BoundingBox, GlyphId, Point},
+    ReadError, TableProvider,
+};
+
+use std::{collections::HashSet, fmt::Debug, ops::Range};
+
+use colrv1_traversal::{get_clipbox_font_units, traverse_with_callbacks};
+
+pub use transforms::Transform;
+
+use crate::prelude::LocationRef;
+
+use self::instance::{resolve_paint, PaintId};
+
+/// An error during drawing a COLR glyph. This covers inconsistencies
+/// in the COLRv1 paint graph as well as downstream
+/// parse errors from read-fonts.
+#[derive(Debug, Clone)]
+pub enum ColorError {
+    ParseError(ReadError),
+    NoColrV1GlyphForId(GlyphId),
+    PaintCycleDetected,
+}
+
+impl std::fmt::Display for ColorError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            ColorError::ParseError(read_error) => {
+                write!(f, "Error parsing font data: {read_error}")
+            }
+            ColorError::NoColrV1GlyphForId(glyph_id) => {
+                write!(f, "No COLRv1 glyph found for glyph id: {glyph_id}")
+            }
+            ColorError::PaintCycleDetected => write!(f, "Paint cycle detected in COLRv1 glyph."),
+        }
+    }
+}
+
+impl From<ReadError> for ColorError {
+    fn from(value: ReadError) -> Self {
+        ColorError::ParseError(value)
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+// This repr(C) is required so that C-side FFI's
+// are able to cast the ColorStop slice to a C-side array pointer.
+#[repr(C)]
+#[derive(PartialEq)]
+#[cfg_attr(test, derive(Serialize, Deserialize))]
+/// A color stop of a gradient. All gradient callbacks of [`ColorPainter`] normalize color stops to be in the range of 0
+/// to 1. `palette_index` specifies a color from the CPAL table, to be multiplied with `alpha` before use.
+pub struct ColorStop {
+    pub offset: f32,
+    pub palette_index: u16,
+    pub alpha: f32,
+}
```

Comment:
```suggestion
/// A color stop of a gradient.
///
/// All gradient callbacks of [`ColorPainter`] normalize color stops to be in the range of 0
/// to 1.
#[derive(Clone, PartialEq, Debug, Default)]
#[cfg_attr(test, derive(Serialize, Deserialize))]
// This repr(C) is required so that C-side FFI's
// are able to cast the ColorStop slice to a C-side array pointer.
#[repr(C)]
pub struct ColorStop {
    pub offset: f32,
    /// Specifies a color from the `CPAL` table.
    pub palette_index: u16,
    /// Additional alpha value, to be multiplied with the color above before use.
    pub alpha: f32,
}
```
Separated the brief comment as above. The general order is doc comment, derives, repr. Also added doc comments to the fields rather than specifying them at the type level.

Reply:
Got it, incorporated suggestion.